ROADMAP.md

```diff
@@ -0,0 +1,192 @@
+- Support multi-build/folder outputs based on ES version/browser/engines?
+- Includes making sure you can run Babel safely over node_modules [like with #6280](https://github.com/babel/babel/pull/6280). [create-react-app has an issue for this](https://github.com/facebookincubator/create-react-app/issues/3777)
+
+### Codemods for TC39 Proposals
+
+> Lebab/others are already used to convert from ES5 -> ESNext, so incorporate it into Babel itself.
+
+- Refactor [Lebab](https://github.com/lebab/lebab/issues/138) as Babel transforms (can keep the cli since it's a separate tool)
+  - Usecase: ES3 -> ES6+ (on source code)
+  - Usecase: Remove usage of dropped proposals
+  - Usecase: Auto upgrade to the latest version of a proposal spec (if possible)
+  - Can we somehow combine forces in: babel-codemod/jscodeshift/lebab, prettier/recast/babel-generator? I really don't want to update all of these: new syntax equals re-writing the printer in all of these places separately/out of sync.
+
+## Increasing the quality of community plugins
+
+- Work with the community to create guides on how to write plugins or understand ASTs, etc
+- Analysis of API's/syntax used (Google BigQuery)
+- Have #blessed/sanctioned/curated packages according to some standard
+  - Can use for smoke tests
+  - Official testing package
+  - Certain level of coverage, downloads, etc
+  - Create a scoped namespace on github/npm for these? like webpack-contrib
+  - Can enforce linting rules on apis?
+  - Makes ecosystem changes easier if can notify and upgrade these plugins beforehand
+  - Create a set of standard tests to verify against (handles all syntax)
+  - Documented/tested set of options
+
+### ASTExplorer
+
+- allow custom version of babel-standalone (same as REPL to allow per PR tests)
+- integrate with repl? (both are in react)
+- auto publish a plugin to npm?
+- create tests?
+- hook into codesandbox?
+
+---
+
+## Feature
+
+### Performance table for ES6+ (used in babel-preset-env under new option e.g `perf`)
+
+> Add a new option in `preset-env`: will continue to compile if the native support is *slower*. Can setup a threshold, may need to compare against the size difference.
+
+- Use [six-speed](https://github.com/kpdecker/six-speed) repo as a base, needs to apply for ES6 and proposals
+- Need continued maintainence
+
+### Compiled Output Stats 
+
+> [#5340](https://github.com/babel/babel/issues/5340) Can show the code size before and after compiling to give a sense of compiled output. Could create suggestions like using "loose" mode or not compiling, etc.
+
+- The [REPL](https://twitter.com/existentialism/status/948940160653291520) just added a before/after code-size
+- Maybe difficult to do per transform
+
+### Async Transforms
+
+Support having async plugins. Will require it to be opt-in and for all other plugins to be async.
+
+### Plugin Ordering
+
+- Add `before`/`after` keys for plugins so they can specify order.
```

Comment:
Encoding before/after within a plugin is insufficient, as a user can compose two plugins entirely unaware of each other. In addition to a plugin being able to specific it's own dependencies, it is inevitable that the user will also need to modify/augment those dependencies.

For example:

A user may have two plugins unaware of each other, but with an implicit order dependency. 

To address this, the user must also have control over the order dependency. One solution is to allow the user to impose an additional edge, to inform the DAG's topsort output.

I have used https://github.com/krisselden/dag-map to solve this problem in the past, and it worked like a charm.

cc @rwjblue  / @krisselden

Reply:
More Thoughts on this:

We may be able to identify some common milestones during the transpilation process, these Milestones can serve as mechanism of grouping plugins together without tightly coupling them.

For example, one may want X category of work to always occur before Y category of work.

Specifically, one typically wants to strip assertions/feature flags, before module transpilation, and never after. This should continue to work regardless of the module transpilation strategy  be it CJS/AMD/magic bundler or even disabled entirely.

In this case, the feature-flag/assert stripping plugin, shouldn't need to be aware of ALL module transpilation plugins, rather it should be able to simply say. I should be able to add an edge, informing the DAG that it must be run before the module transpilation miletones.

Reply:
Sorry I didn't link to the older prs maybe but I was suggesting allowing a plugin to opt into a "capability" so it wouldn't be hardcoded to a plugin name - like for modules or anything else, I believe that's the same thought you had with a category

Reply:
Sorry I didn't link to the older prs maybe but I was suggesting allowing a plugin to opt into a "capability" so it wouldn't be hardcoded to a plugin name - like for modules or anything else, I believe that's the same thought you had with a category

Reply:
@stefanpenner maybe babel could at least log warnings for plugins that know they're in the wrong order relative to each other though?  For instance it sounds doubtful that there will ever be a valid use case for putting `proposal-decorators` after `proposal-class-properties`, so a warning there would help a lot of people, especially since it's not at all obvious what's wrong from the errors that can result.